biorxiv/pca_association_experiment/04_biorxiv_word_pca_component_doc2vec.py

```diff
@@ -0,0 +1,194 @@
+# ---
+# jupyter:
+#   jupytext:
+#     formats: ipynb,py
+#     text_representation:
+#       extension: .py
+#       format_name: light
+#       format_version: '1.5'
+#       jupytext_version: 1.9.1+dev
+#   kernelspec:
+#     display_name: Python [conda env:annorxiver]
+#     language: python
+#     name: conda-env-annorxiver-py
+# ---
+
+# # Determine Word to PCA Associations
+
+# This notebook is designed to run PCA over the document embeddings and calculate words-pca associations and document centroid-pca associations for each principal component.
+
+# +
+from pathlib import Path
+import os
+import re
+
+from gensim.models import Doc2Vec
+import itertools
+import matplotlib
+import matplotlib.pyplot as plt
+import pandas as pd
+import plotnine as p9
+from PIL import ImageColor
+from scipy.spatial.distance import cdist
+from sklearn.decomposition import PCA
+from tqdm import tqdm_notebook
+import wordcloud
+
+matplotlib.use("SVG")  # set the backend to SVG
+# -
+
+journal_map_df = pd.read_csv(
+    "../exploratory_data_analysis/output/biorxiv_article_metadata.tsv", sep="\t"
+)
+journal_map_df.head()
+
+# # Get the Word Vectors
+
+# Save the word vectors to disk, so later sections have easy access.
+
+model = Doc2Vec.load(
+    "../word_vector_experiment/output/doc2vec_output/biorxiv_300.model"
+)
+
+# # PCA the Documents
+
+# Run PCA over the documents. Generates 50 principal components, but can generate more or less.
+
+n_components = 50
+random_state = 100
+
+# +
+reducer = PCA(n_components=n_components, random_state=random_state)
+
+embedding = reducer.fit_transform(model.docvecs.vectors_docs)
+
+pca_df = (
+    pd.DataFrame(
+        embedding, columns=[f"pca{dim}" for dim in range(1, n_components + 1, 1)]
+    )
+    .assign(document=[f"{str(tag)}.xml" for tag in model.docvecs.doctags])
+    .merge(journal_map_df[["category", "document", "doi"]], on="document")
+)
+# -
+
+(
+    pd.DataFrame(
+        reducer.components_,
+        columns=[f"{dim+1}" for dim in range(reducer.components_.shape[1])],
+    ).to_csv(
+        "output/word_pca_similarity/pca_components_doc2vec.tsv", sep="\t", index=False
+    )
+)
+
+# # Calculate Word-PCA Cosine Similarity
+
+# Once PCA has finished, there are now 50 different principal components. The association between every word and principal component is calculated via [cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity) (cosine of the angle between two vectors).
+
+# +
+# 1 - cosine distance = cosine similarity
+word_pca_similarity = 1 - cdist(model.wv.vectors, reducer.components_, "cosine")
+
+word_pca_similarity.shape
```

Comment:
So here `model.wv.vectors` has word vectors in rows and documents in columns? I understand these document vectors are your averages across word vectors. And `reducer.components_` has vectors in rows and, well, the principal components (combinations of document vectors)? `word_pca_similarity` has the similarity of the words vectors in different representations (documents vs PCA documents). Sorry, just thinking aloud to check from you if it makes sense.

Reply:
>So here model.wv.vectors has word vectors in rows and documents in columns? I understand these document vectors are your averages across word vectors. And reducer.components_ has vectors in rows and, well, the principal components (combinations of document vectors)? word_pca_similarity has the similarity of the words vectors in different representations (documents vs PCA documents). Sorry, just thinking aloud to check from you if it makes sense.

Close. Gensim uses the convention 'wv' to refer to word vectors, so in this case I'm calculating the cosine distance between word vectors (rows in `model.wv.vectors`) and the principal components (rows in `reducer.components_`). These principal components are generated by running PCA on the document vectors generated by doc2vec.  So `word_pca_similarity` is a similarity matrix of words as rows (also generated by the same doc2vec model), PC components as columns and the similarity score as entries.